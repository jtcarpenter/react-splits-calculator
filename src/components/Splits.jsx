import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import raceConfig from '../config/raceConfig';

const A4_PAPER_LENGTH_MM = 297;
const PRINT_MARGIN_HEIGHT = 10;
const PRINT_THEAD_HEIGHT = 10;
const PRINT_FONT_SIZE_SCALE = 1;
const PRINT_BORDER_WIDTH = .5;
const PRINT_WRIST_SIZE = 150;
const MAX_PRINT_ROW_HEIGHT = 10;
const PRINT_TFOOT_HEIGHT =  A4_PAPER_LENGTH_MM -
                            PRINT_WRIST_SIZE -
                            PRINT_THEAD_HEIGHT -
                            (PRINT_MARGIN_HEIGHT * 2);

function resolvePrintRowHeight(numSplits) {
    const calculatedPrintRowHeight = (
        (PRINT_WRIST_SIZE - PRINT_THEAD_HEIGHT) / parseInt(numSplits + 1)
    );
    return calculatedPrintRowHeight < MAX_PRINT_ROW_HEIGHT
        ? calculatedPrintRowHeight
        : MAX_PRINT_ROW_HEIGHT;
}

function resolvePrintRowFontSize(resolvePrintRowHeight) {
    return resolvePrintRowHeight * PRINT_FONT_SIZE_SCALE;
}

const StyledTable = styled.table`
    page-break-inside: avoid;
    table-layout:fixed
    margin: 0;
    padding: 0;
    border-collapse: collapse;
    border-spacing: 0;
    border: ${PRINT_BORDER_WIDTH}px dashed ${(props) => props.theme.DIVIDER};
    float: left;

    @media print {
        border-width: ${PRINT_BORDER_WIDTH}mm;
        margin: ${PRINT_MARGIN_HEIGHT}mm;
    }
`;

const StyledTHeadTR = styled.tr`
    display: table-row;
    border-collapse: collapse;
    border-spacing: 0;
    margin: 0;
    border-bottom: ${PRINT_BORDER_WIDTH}px solid ${(props) => props.theme.DIVIDER};

    @media print {
        border-bottom-width: ${PRINT_BORDER_WIDTH}mm;
        height: ${(props) => { return props.printRowHeight; }}mm;
        font-size: ${(props) => { return props.printRowFontSize; }}mm;
    }
`;

const StyledTR = styled.tr`
    display: table-row;
    margin: 0;
    border-bottom: ${PRINT_BORDER_WIDTH}px solid ${(props) => props.theme.DIVIDER};

    &:last-child {
        border-bottom: ${PRINT_BORDER_WIDTH}px dashed ${(props) => props.theme.DIVIDER};
    };

    @media print {
        border-bottom-width: ${PRINT_BORDER_WIDTH}mm;
        height: ${(props) => { return props.printRowHeight; }}mm;
        font-size: ${(props) => { return props.printRowFontSize; }}mm;
        &:last-child {
            border-bottom: ${PRINT_BORDER_WIDTH}mm dashed ${(props) => props.theme.DIVIDER};
        }
    }
`;

const StyledTFootTR = styled.tr`
    display: none;
    padding: 0;
    margin: 0;
    border: none;

    @media print {
        display: table-row;
        height: ${PRINT_TFOOT_HEIGHT}mm;
    }
`;

const StyledTH = styled.th`
    line-height: 1;
`;

const StyledTD = styled.td`
    text-align: center;
    padding: 0 10px;
    margin: 0;
    line-height: 1;
    table-layout:fixed;

    &:first-child {
        padding-right: 5px;
    };
    &:last-child {
        padding-left: 5px;
        font-weight: bold;
    };

    @media print {
        padding: 0 2mm;
    }
`;

const StyledOl = styled.ul`
    display: none;
    list-style: none;
    padding: 0;
    margin: 0;
    float: left;

    @media print {
        display: block;
        margin-top: ${PRINT_MARGIN_HEIGHT}mm;
    }
`;

const StyledLi = styled.li`
    display: block;
    height: 9.5mm;
    position: relative;

    &:first-child {
        height: 9mm;
        border-top: .5mm solid black;
    }
`;

const StyledSpan = styled.span`
    position: absolute;
    bottom: 0;
    border-bottom: .5mm solid black;
`;

const StyledP = styled.p`
    display: block;
    float: left;
`;

export function Splits({ splits, raceId, raceUnit }) {

    function parseSplit(split) {
        let time = split.time;
        const hours = `${Math.floor(time / (60*60))}`;
        time = time % (60*60);
        const minutes = `0${Math.floor(time / 60)}`.substr(-2, 2);
        time = time % 60;
        const secs = `0${Math.floor(time)}`.substr(-2, 2);
        return`${hours}:${minutes}:${secs}`;
    }

    function raceUnitName(raceUnitId) {
        const raceUnit =  raceConfig.raceUnits.find((raceUnit) => {
            return raceUnit.id === raceUnitId;
        });
        return raceUnit ? raceUnit.name : '';
    }

    function parseSplitNumber(split, raceId, raceUnit) {
        if (split.last) {
            return Math.round(raceConfig.races.find((raceUnit) => {
                return raceUnit.id === raceId;
            })[raceUnit] * 10) / 10;
        }
        return split.number;
    }

    const printRowHeight = resolvePrintRowHeight(splits.length);
    const printRowFontSize = resolvePrintRowFontSize(printRowHeight);
    const numMarkers = Math.floor((A4_PAPER_LENGTH_MM - PRINT_MARGIN_HEIGHT) / 10);
    const markers = Array.apply(null, Array(numMarkers)).map((v, i) => i + 1);

    return (
        <Fragment>
            <StyledTable>
                <thead>
                    <StyledTHeadTR
                        printRowHeight={ printRowHeight }
                        printRowFontSize={ printRowFontSize }
                    >
                        <StyledTH>{ raceUnitName(raceUnit) }</StyledTH>
                        <StyledTH>Split</StyledTH>
                    </StyledTHeadTR>
                </thead>
                <tbody>
                    {splits.map((split, index) =>
                        <StyledTR
                            key={ index }
                            printRowHeight= { printRowHeight }
                            printRowFontSize={ printRowFontSize }
                        >
                            <StyledTD>
                                { parseSplitNumber(split, raceId, raceUnit) }
                            </StyledTD>
                            <StyledTD>
                                { parseSplit(split) }
                            </StyledTD>
                        </StyledTR>
                    )}
                </tbody>
                <tfoot>
                    <StyledTFootTR>
                        <StyledTD colSpan="2"></StyledTD>
                    </StyledTFootTR>
                </tfoot>
            </StyledTable>
            <StyledOl>
                {markers.map((mark, index) =>
                    <StyledLi key={ index }>
                        <StyledSpan>{ mark }CM</StyledSpan>
                    </StyledLi>
                )}
            </StyledOl>
            <StyledP>Instructions</StyledP>
        </Fragment>
    );
}

Splits.propTypes = {
    splits: PropTypes.array.isRequired,
    raceUnit: PropTypes.string.isRequired,
    raceId: PropTypes.string.isRequired
};

export default Splits;