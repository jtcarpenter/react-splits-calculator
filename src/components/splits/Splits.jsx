import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import raceConfig from '../../config/raceConfig';
import Table from './Table.jsx';
import TRHead from './TRHead.jsx';
import TR from './TR.jsx';
import TRFoot from './TRFoot.jsx';
import TH from './TH.jsx';
import TD from './TD.jsx';
import OL from './OL.jsx';
import LI from './LI.jsx';
import Span from './Span.jsx';

const A4_PAPER_LENGTH_MM = 297;
const PRINT_MARGIN_HEIGHT = 10;
const PRINT_THEAD_HEIGHT = 10;
const PRINT_FONT_SIZE_SCALE = 1;
const PRINT_BORDER_WIDTH = .5;
const PRINT_WRIST_SIZE = 150;
const MAX_PRINT_ROW_HEIGHT = 10;
const PRINT_TFOOT_HEIGHT =  A4_PAPER_LENGTH_MM -
                            PRINT_WRIST_SIZE -
                            PRINT_THEAD_HEIGHT -
                            (PRINT_MARGIN_HEIGHT * 2);

function resolvePrintRowHeight(numSplits) {
    const calculatedPrintRowHeight = (
        (PRINT_WRIST_SIZE - PRINT_THEAD_HEIGHT) / parseInt(numSplits + 1)
    );
    return calculatedPrintRowHeight < MAX_PRINT_ROW_HEIGHT
        ? calculatedPrintRowHeight
        : MAX_PRINT_ROW_HEIGHT;
}

function resolvePrintRowFontSize(resolvePrintRowHeight) {
    return resolvePrintRowHeight * PRINT_FONT_SIZE_SCALE;
}

export function Splits({ splits, raceId, raceUnit }) {

    function parseSplit(split) {
        let time = split.time;
        const hours = `${Math.floor(time / (60*60))}`;
        time = time % (60*60);
        const minutes = `0${Math.floor(time / 60)}`.substr(-2, 2);
        time = time % 60;
        const secs = `0${Math.floor(time)}`.substr(-2, 2);
        return`${hours}:${minutes}:${secs}`;
    }

    function raceUnitName(raceUnitId) {
        const raceUnit =  raceConfig.raceUnits.find((raceUnit) => {
            return raceUnit.id === raceUnitId;
        });
        return raceUnit ? raceUnit.name : '';
    }

    function parseSplitNumber(split, raceId, raceUnit) {
        if (split.last) {
            return Math.round(raceConfig.races.find((raceUnit) => {
                return raceUnit.id === raceId;
            })[raceUnit] * 10) / 10;
        }
        return split.number;
    }

    const printRowHeight = resolvePrintRowHeight(splits.length);
    const printRowFontSize = resolvePrintRowFontSize(printRowHeight);
    const numMarkers = Math.floor((A4_PAPER_LENGTH_MM - PRINT_MARGIN_HEIGHT) / 10);
    const markers = Array.apply(null, Array(numMarkers)).map((v, i) => i + 1);

    return (
        <Fragment>
            <Table
                printBorderWidth={ PRINT_BORDER_WIDTH }
                printMarginHeight={ PRINT_MARGIN_HEIGHT }
            >
                <thead>
                    <TRHead
                        printRowHeight={ printRowHeight }
                        printRowFontSize={ printRowFontSize }
                        printBorderWidth={ PRINT_BORDER_WIDTH }
                    >
                        <TH>{ raceUnitName(raceUnit) }</TH>
                        <TH>Split</TH>
                    </TRHead>
                </thead>
                <tbody>
                    {splits.map((split, index) =>
                        <TR
                            key={ index }
                            printRowHeight= { printRowHeight }
                            printRowFontSize={ printRowFontSize }
                            printBorderWidth={ PRINT_BORDER_WIDTH }
                        >
                            <TD>
                                { parseSplitNumber(split, raceId, raceUnit) }
                            </TD>
                            <TD>
                                { parseSplit(split) }
                            </TD>
                        </TR>
                    )}
                </tbody>
                <tfoot>
                    <TRFoot printTFootHeight={ PRINT_TFOOT_HEIGHT }>
                        <TD colSpan="2"></TD>
                    </TRFoot>
                </tfoot>
            </Table>
            <OL printMarginHeight={ PRINT_MARGIN_HEIGHT }>
                {markers.map((mark, index) =>
                    <LI key={ index }>
                        <Span>{ mark }CM</Span>
                    </LI>
                )}
            </OL>
        </Fragment>
    );
}

Splits.propTypes = {
    splits: PropTypes.array.isRequired,
    raceUnit: PropTypes.string.isRequired,
    raceId: PropTypes.string.isRequired
};

export default Splits;