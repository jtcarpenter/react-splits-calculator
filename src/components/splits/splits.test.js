import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { Splits } from './Splits.jsx';

const splits = [];
const raceUnit = '';
const raceId = '';

const splitsComp = (
    <Splits
        splits={ splits }
        raceUnit={ raceUnit }
        raceId={ raceId }
    ></Splits>
);

describe('Splits', () => {
    it('should render', () => {
        const wrapper = shallow(splitsComp);
        expect(wrapper.exists()).toBe(true);
    });

    it('should display correctly', () => {
        const splits = renderer.create(splitsComp);
        expect(splits.toJSON()).toMatchSnapshot();
    });
});