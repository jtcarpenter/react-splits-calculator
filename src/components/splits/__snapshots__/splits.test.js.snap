// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Splits should display correctly 1`] = `
<div
  className="sc-dnqmqq irAkrN"
>
  <table
    className="sc-bdVaJa kxxYvM"
  >
    <thead>
      <tr
        className="sc-bwzfXH WNlQL"
      >
        <th
          className="sc-ifAKCX eSDSEW"
        >
          
        </th>
        <th
          className="sc-ifAKCX eSDSEW"
        >
          Split
        </th>
      </tr>
    </thead>
    <tbody />
    <tfoot>
      <tr
        className="sc-bxivhb jUTdPV"
      >
        <td
          className="sc-EHOje kFKlGs"
          colSpan="2"
        />
      </tr>
    </tfoot>
  </table>
  <ol
    className="sc-bZQynM jLHPZU"
  >
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        1
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        2
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        3
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        4
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        5
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        6
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        7
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        8
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        9
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        10
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        11
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        12
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        13
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        14
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        15
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        16
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        17
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        18
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        19
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        20
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        21
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        22
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        23
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        24
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        25
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        26
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        27
        CM
      </span>
    </li>
    <li
      className="sc-gzVnrw kIuHPV"
    >
      <span
        className="sc-htoDjs daqvur"
      >
        28
        CM
      </span>
    </li>
  </ol>
</div>
`;
